vernac/opaques.ml

```diff
@@ -0,0 +1,160 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+(************************************************************************)
+(** {6 Tables of opaque proof terms} *)
+
+(** We now store opaque proof terms apart from the rest of the environment.
+    This way, we can quickly load a first half of a .vo file without these opaque
+    terms, and access them only when a specific command (e.g. Print or
+    Print Assumptions) needs it. *)
+
+type 'a const_entry_body = 'a Entries.proof_output Future.computation
+
+type opaque_result =
+| OpaqueCertif of Safe_typing.opaque_certificate Future.computation
+| OpaqueValue of Opaqueproof.opaque_proofterm
+
+(** Current table of opaque terms *)
+
+module Summary =
+struct
+  type t = opaque_result Opaqueproof.HandleMap.t
+  let state : t ref = ref Opaqueproof.HandleMap.empty
+  let init () = state := Opaqueproof.HandleMap.empty
+  let freeze ~marshallable =
+    if marshallable then
+      let iter _ pf = match pf with
+      | OpaqueCertif cert -> ignore (Future.join cert)
+      | OpaqueValue _ -> ()
+      in
+      let () = Opaqueproof.HandleMap.iter iter !state in
+      !state
+    else !state
+  let unfreeze s = state := s
+
+  let join ?(except=Future.UUIDSet.empty) () =
+    let iter i pf = match pf with
+    | OpaqueValue _ -> ()
+    | OpaqueCertif cert ->
+      if Future.UUIDSet.mem (Future.uuid cert) except then ()
+      else if Safe_typing.is_filled_opaque i (Global.safe_env ()) then
+        assert (Future.is_over cert)
+      else
+        (* Little belly dance to preserve the fix_exn wrapper around filling *)
+        let () = Future.join @@ Future.chain cert (fun cert -> Global.fill_opaque cert) in
```

Comment:
The eta expansion is not needed, my bad I suggested it.

Reply:
I like it though, it makes the code more robust to stuff like optional arguments.